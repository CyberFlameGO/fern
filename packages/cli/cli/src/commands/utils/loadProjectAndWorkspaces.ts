import { loadProjectConfig, ProjectConfig } from "@fern-api/project-configuration";
import { getUniqueWorkspaces } from "./getUniqueWorkspaces";

export async function loadProjectAndWorkspaces({
    commandLineWorkspaces,
}: {
    commandLineWorkspaces: readonly string[];
}): Promise<{
    projectConfig: ProjectConfig;
    workspacePaths: string[];
}> {
    const projectConfig = await loadProjectConfig();

    const workspacePaths = await getUniqueWorkspaces({
        commandLineWorkspaces,
        projectConfig,
    });

    return { projectConfig, workspacePaths };
}
