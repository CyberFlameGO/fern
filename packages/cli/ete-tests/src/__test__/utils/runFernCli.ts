import execa, { Options } from "execa";
import path from "path";

export async function runFernCli(args: string[], options?: Options): Promise<execa.ExecaChildProcess> {
    const cmd = execa("node", [path.join(__dirname, "../../../../cli/dist/bundle.cjs"), ...args], options);
    cmd.stdout?.pipe(process.stdout);
    cmd.stderr?.pipe(process.stderr);
    return cmd;
}
