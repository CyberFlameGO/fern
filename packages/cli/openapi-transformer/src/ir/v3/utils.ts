import { FernOpenapiIr } from "@fern-fern/openapi-ir-sdk";
import { OpenAPIV3 } from "openapi-types";

export const APPLICATION_JSON_CONTENT = "application/json";
export const SCHEMA_REFERENCE_PREFIX = "#/components/schemas/";
export const RESPONSE_REFERENCE_PREFIX = "#/components/responses/";
export const REQUEST_REFERENCE_PREFIX = "#/components/requests/";

export function isReferenceObject(
    parameter:
        | OpenAPIV3.ReferenceObject
        | OpenAPIV3.ParameterObject
        | OpenAPIV3.SchemaObject
        | OpenAPIV3.RequestBodyObject
        | OpenAPIV3.SecuritySchemeObject
): parameter is OpenAPIV3.ReferenceObject {
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    return (parameter as OpenAPIV3.ReferenceObject).$ref != null;
}

export function getSchemaIdFromReference(reference: OpenAPIV3.ReferenceObject): FernOpenapiIr.SchemaId | undefined {
    if (reference.$ref.startsWith(SCHEMA_REFERENCE_PREFIX)) {
        return reference.$ref.substring(SCHEMA_REFERENCE_PREFIX.length);
    }
    return undefined;
}
