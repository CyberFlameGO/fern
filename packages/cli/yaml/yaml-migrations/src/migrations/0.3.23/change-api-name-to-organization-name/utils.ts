import { AbsoluteFilePath, join } from "@fern-api/fs-utils";
import { TaskContext } from "@fern-api/task-context";
import { findUp } from "find-up";
import { readFile } from "fs/promises";
import glob from "glob-promise";

const FERN_DIRECTORY = "fern";
const FERN_ROOT_CONFIG_FILENAME = "fern.config.json";

export async function getOrganizationName(context: TaskContext): Promise<string> {
    const fernDirectory = await getFernDirectory();
    if (fernDirectory == null) {
        context.failAndThrow(`Directory "${FERN_DIRECTORY}" not found.`);
    }
    const fernConfigJsonBuffer = await readFile(join(fernDirectory, FERN_ROOT_CONFIG_FILENAME));
    const fernConfigJson = JSON.parse(fernConfigJsonBuffer.toString());
    const organizationName = fernConfigJson.organization;
    if (organizationName == null) {
        context.failAndThrow(`${FERN_ROOT_CONFIG_FILENAME} is missing organization`);
    }
    return organizationName;
}

export async function getAllApiYmlFiles(context: TaskContext): Promise<AbsoluteFilePath[]> {
    const fernDirectory = await getFernDirectory();
    if (fernDirectory == null) {
        context.failAndThrow(`Directory "${FERN_DIRECTORY}" not found.`);
    }
    const filepaths = await glob("*/definition/api.yml", {
        cwd: fernDirectory,
        absolute: true,
    });
    return filepaths.map(AbsoluteFilePath.of);
}

async function getFernDirectory(): Promise<AbsoluteFilePath | undefined> {
    const fernDirectoryStr = await findUp(FERN_DIRECTORY, { type: "directory" });
    if (fernDirectoryStr == null) {
        return undefined;
    }
    return AbsoluteFilePath.of(fernDirectoryStr);
}
