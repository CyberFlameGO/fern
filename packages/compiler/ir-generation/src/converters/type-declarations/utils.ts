import { RawSchemas } from "@fern-api/syntax-analysis";

export function isRawAliasDefinition(
    rawTypeDeclaration: RawSchemas.TypeDeclarationSchema
): rawTypeDeclaration is RawSchemas.AliasSchema {
    return (rawTypeDeclaration as RawSchemas.AliasSchema)?.alias != null;
}

export function isRawObjectDefinition(
    rawTypeDeclaration: RawSchemas.TypeDeclarationSchema
): rawTypeDeclaration is RawSchemas.ObjectSchema {
    return (rawTypeDeclaration as RawSchemas.ObjectSchema)?.properties != null;
}

export function isRawUnionDefinition(
    rawTypeDeclaration: RawSchemas.TypeDeclarationSchema
): rawTypeDeclaration is RawSchemas.UnionSchema {
    return (rawTypeDeclaration as RawSchemas.UnionSchema)?.union != null;
}

export function isRawEnumDefinition(
    rawTypeDeclaration: RawSchemas.TypeDeclarationSchema
): rawTypeDeclaration is RawSchemas.EnumSchema {
    return (rawTypeDeclaration as RawSchemas.EnumSchema)?.enum != null;
}
