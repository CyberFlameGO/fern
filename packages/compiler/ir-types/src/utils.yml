# These types aren't actually present in the IR, but may be helpful
# for generators

imports:
  types: ./types.yml
  errors: ./errors.yml

types:
  ModelReference:
    union:
      type: types.DeclaredTypeName
      error: errors.ErrorName
