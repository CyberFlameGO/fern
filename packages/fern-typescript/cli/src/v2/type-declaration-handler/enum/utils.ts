import { EnumValue } from "@fern-fern/ir-model";
import snakeCase from "lodash-es/snakeCase";
import upperFirst from "lodash-es/upperFirst";

export function getKeyForEnum({ name }: EnumValue): string {
    return snakeCase(name).split("_").map(upperFirst).join("");
}

export function getValueForEnum({ value }: EnumValue): string {
    return snakeCase(value).toUpperCase();
}
