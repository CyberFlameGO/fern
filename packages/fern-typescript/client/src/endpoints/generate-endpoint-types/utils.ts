import { NamedType } from "@fern-api/api";
import { generateTypeReference } from "@fern-typescript/commons";
import { Directory, SourceFile, ts } from "ts-morph";
import { generateEndpointTypeReference } from "../generateEndpointTypeReference";
import { WireMessageBodyReference } from "./types";

// adds an import to the `referencedIn` file and returns a reference to the imported type
export function generateReferenceToWireMessageType({
    serviceName,
    endpointId,
    servicesDirectory,
    reference,
    referencedIn,
    modelDirectory,
}: {
    serviceName: NamedType;
    endpointId: string;
    referencedIn: SourceFile;
    modelDirectory: Directory;
    servicesDirectory: Directory;
    reference: WireMessageBodyReference;
}): ts.TypeNode {
    if (reference.isLocal) {
        return generateEndpointTypeReference({
            serviceName,
            endpointId,
            typeName: reference.typeName,
            referencedIn,
            servicesDirectory,
        });
    } else {
        return generateTypeReference({
            reference: reference.typeReference,
            referencedIn,
            modelDirectory,
        });
    }
}
