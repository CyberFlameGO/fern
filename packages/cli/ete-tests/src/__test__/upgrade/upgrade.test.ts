import { WORKSPACE_CONFIGURATION_FILENAME } from "@fern-api/workspace-configuration";
import { readFile } from "fs/promises";
import path from "path";
import { runFernCli } from "../utils/runFernCli";

const FIXTURES_DIR = path.join(__dirname, "fixtures");

describe("fern-upgrade-tests", () => {
    itFixture("simple");
});

function itFixture(fixtureName: string) {
    it(
        // eslint-disable-next-line jest/valid-title
        fixtureName,
        async () => {
            const fixturePath = path.join(FIXTURES_DIR, fixtureName);

            await runFernCli(["upgrade"], {
                cwd: fixturePath,
                reject: false,
            });

            const fileContents = await readFile(path.join(fixturePath, "api", WORKSPACE_CONFIGURATION_FILENAME));
            expect(fileContents.toString()).toMatchSnapshot();
        },
        90_000
    );
}
