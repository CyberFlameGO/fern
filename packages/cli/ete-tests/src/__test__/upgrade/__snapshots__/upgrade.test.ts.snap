// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fern-ir-tests simple 1`] = `
"[31mType [1mMissingType[22m is not defined.[39m
  [34mapi.yml -> types -> MyType[39m
  "
`;

exports[`fern-upgrade-tests simple 1`] = `
"name: api
definition: src
generators:
  - name: fernapi/fern-postman
    version: 0.0.21
    generate:
      enabled: true
      output: ./generated-postman.json
  - name: fernapi/fern-openapi
    version: 0.0.4
    generate:
      enabled: true
      output: ./generated-openapi.yml
    config:
      format: yaml
  - name: fernapi/fern-java
    version: 0.0.84
    generate: true
    config:
      packagePrefix: com
      mode: client_and_server
  - name: fernapi/fern-typescript
    version: 0.0.155
    generate: true
    config:
      mode: client_and_server
"
`;
